server/sonar-server/src/main/java/org/sonar/server/computation/formula/AverageCounter.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.formula;
+
+import com.google.common.base.Optional;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.measure.Measure;
+
+public class AverageCounter implements Counter<AverageCounter> {
+
+  private final String mainMetric;
+  private final String byMetric;
+  @CheckForNull
+  private final String fallbackMetric;
+
+  private Double mainValue;
+  private Double byValue;
+
+  public AverageCounter(String mainMetric, String byMetric, @Nullable String fallbackMetric) {
+    this.byMetric = byMetric;
+    this.mainMetric = mainMetric;
+    this.fallbackMetric = fallbackMetric;
+  }
+
+  @Override
+  public void aggregate(AverageCounter counter) {
+    addValuesIfPresent(counter.getMainValue(), counter.getByValue());
+  }
+
+  @Override
+  public void aggregate(Component component, CounterContext counterContext) {
+    Optional<Double> mainValueOptional = getDoubleValue(counterContext.getMeasure(mainMetric));
+    if (!mainValueOptional.isPresent() && fallbackMetric != null) {
+      mainValueOptional = getDoubleValue(counterContext.getMeasure(fallbackMetric));
+    }
+    Optional<Double> byValueOptional = getDoubleValue(counterContext.getMeasure(byMetric));
+    addValuesIfPresent(mainValueOptional, byValueOptional);
+  }
+
+  private void addValuesIfPresent(Optional<Double> counterMainValue, Optional<Double> counterByValue){
+    if (counterMainValue.isPresent() && counterByValue.isPresent()) {
+      mainValue = mainValue == null ? counterMainValue.get() : mainValue + counterMainValue.get();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)